cpp/include/cudf/table/element_equality_comparator.cuh

```diff
@@ -0,0 +1,120 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/table/table_device_view.cuh>
+
+namespace cudf {
+/**
+ * @brief A specialization for floating-point `Element` type to check if
+ * `lhs` is equivalent to `rhs`. `nan == nan`.
+ *
+ * @param[in] lhs first element
+ * @param[in] rhs second element
+ * @return bool `true` if `lhs` == `rhs` else `false`.
+ */
+template <typename Element, std::enable_if_t<std::is_floating_point<Element>::value>* = nullptr>
+__device__ bool equality_compare(Element lhs, Element rhs)
+{
+  if (isnan(lhs) and isnan(rhs)) { return true; }
+  return lhs == rhs;
+}
+
+/**
+ * @brief A specialization for non-floating-point `Element` type to check if
+ * `lhs` is equivalent to `rhs`.
+ *
+ * @param[in] lhs first element
+ * @param[in] rhs second element
+ * @return bool `true` if `lhs` == `rhs` else `false`.
+ */
+template <typename Element, std::enable_if_t<not std::is_floating_point<Element>::value>* = nullptr>
+__device__ bool equality_compare(Element const lhs, Element const rhs)
+{
+  return lhs == rhs;
+}
+
+/**
+ * @brief Performs an equality comparison between two elements in two columns.
+ *
+ * @tparam has_nulls Indicates the potential for null values in either column.
+ **/
+template <bool has_nulls = true>
+class element_equality_comparator {
+ public:
+  /**
+   * @brief Construct type-dispatched function object for comparing equality
+   * between two elements.
+   *
+   * @note `lhs` and `rhs` may be the same.
+   *
+   * @param lhs The column containing the first element
+   * @param rhs The column containing the second element (may be the same as lhs)
+   * @param nulls_are_equal Indicates if two null elements are treated as equivalent
+   **/
+  __host__ __device__ element_equality_comparator(column_device_view lhs,
+                                                  column_device_view rhs,
+                                                  bool nulls_are_equal = true)
+    : lhs{lhs}, rhs{rhs}, nulls_are_equal{nulls_are_equal}
+  {
+  }
+
+  /**
+   * @brief Compares the specified elements for equality.
+   *
+   * @param lhs_element_index The index of the first element
+   * @param rhs_element_index The index of the second element
+   *
+   */
+  template <typename Element,
+            std::enable_if_t<cudf::is_equality_comparable<Element, Element>()>* = nullptr>
+  __device__ bool operator()(size_type lhs_element_index, size_type rhs_element_index) const
+    noexcept
+  {
+    if (has_nulls) {
+      bool const lhs_is_null{lhs.nullable() and lhs.is_null(lhs_element_index)};
+      bool const rhs_is_null{rhs.nullable() and rhs.is_null(rhs_element_index)};
```

Comment:
While we're at it, might as well clean this up. This nullability check is redundant. 
```suggestion
      bool const lhs_is_null{lhs.is_null(lhs_element_index)};
      bool const rhs_is_null{rhs.is_null(rhs_element_index)};
```